libpod/image/image.go

```diff
@@ -847,6 +847,23 @@ func (i *Image) Dangling() bool {
 	return len(i.Names()) == 0
 }
 
+// IsCacheImage returns true if the image is cache or intermediate image.
+// Cache image has parent and child.
```

Comment:
I am note 100 percent sure about that. @mtrmac @nalind , WDYT?

Reply:
First, the function is misnamed. Let’s not confuse ourselves like that.

---

The bug report says
> Basically, docker image prune does not remove any images which have dependent tagged images.

This is clearly not implementing _that_, e.g. a chain of 5 derived dangling images will AFAICS cause only the most-derived image to be pruned, when all 5 should be pruned.

---

(I would also expect an O(N) mark&sweep GC over the layer tree (we are essentially pruning untagged _layers_, leaving any attached image objects around, after all, and relative parent/child relationships for the same `TopLayer` don’t really matter), this is O(N^2). Or, at the _very least_, have this `IsCacheImage` bail out early if `!parent` instead of unnecessarily computing `GetParent`.)

Reply:
> This is clearly not implementing that, e.g. a chain of 5 derived dangling images will AFAICS cause only the most-derived image to be pruned, when all 5 should be pruned.

This is not clear to me.
The bug is `podman images prune` should not prune any image( tagged or untagged) which is used as an intermediate image for building finally a tagged image, So while re-build process of the tagged image, all intermediate images should be readily used as cache image.

How to find an image is an intermediate/cache image.
-  image must have child and parent image.
 - This is implemented in `IsCacheImage()` function.

How will the (chain of ) dangling images will be handled?
- If an image has no label and has no child (image derived from this) will be deleted.
- dangling image is formed when the image rebuild happens multiple times, and it leads to the previous target image untagged and not used.
- So dangling images will not have chain of dangling images. The leaf node of the image tree will be dangling.

The above implementation is based on the above points.

Reply:
I think you are saying the same thing: to delete a chain of dangling images, one has to run the proposed implementation N times, where N is the length of the chain.

Reply:
…and as to how a chain of dangling images can be created, consider a change to one of the early lines of a Dockerfile, not the last line. At the point that early line is processed, a new intermediate image is created, and all following images are children of that new intermediate image. The previous intermediate image, and all its children, will be dangling, as a result of that one build.

Reply:
thanks for explanation, got your point.

Those chain of dangling or dependent images used only by dangling images get deleted along with deletion of dangling image(leaf node). This is taken care by `image.Remove()` function which is used for removal of image.
https://github.com/containers/libpod/blob/e1ffac6cc73eb36640cbaf6a1a28ba44749a96d9/libpod/image/image.go#L426 
It deletes all child images which are not base image for any other image. https://github.com/containers/libpod/blob/e1ffac6cc73eb36640cbaf6a1a28ba44749a96d9/libpod/image/image.go#L444

Reply:
Oh, I didn’t realize the list generated by GetPruneImages is not the only images that are deleted. OK, then this probably does what’s necessary.


(I still object to naming a function that returns `true` not only on cache images `IsCache`, that’s going to horribly confuse someone eventually.)

Reply:
> (I still object to naming a function that returns true not only on cache images IsCache, that’s going to horribly confuse someone eventually.)

Will `image.Intermediate` or `image.HasParentAndChild` will work?

Reply:
Both work for me.